File: Back UP/Pico_code/mcp9600.py
from machine import I2C
import struct

_DEFAULT_ADDRESS = 0x67

_REGISTER_HOT_JUNCTION = 0x00
_REGISTER_DELTA_TEMP = 0x01
_REGISTER_COLD_JUNCTION = 0x02
_REGISTER_THERM_CFG = 0x05
_REGISTER_VERSION = 0x20


class MCP9600:

    TYPES = ("K", "J", "T", "N", "S", "E", "B", "R")

    def __init__(self, i2c: I2C, address=_DEFAULT_ADDRESS, tctype="K", tcfilter=0):
        self.i2c = i2c
        self.address = address

        if tctype not in self.TYPES:
            raise ValueError("Invalid thermocouple type")

        tcfilter = max(0, min(7, tcfilter))
        ttype = self.TYPES.index(tctype)

        # Configure thermocouple type + filter
        config_value = tcfilter | (ttype << 4)
        self._write_register(_REGISTER_THERM_CFG, bytes([config_value]))

        # Verify device ID
        device_id = self._read_device_id()
        if device_id not in (0x40, 0x41):
            raise RuntimeError("MCP9600 not found. Check wiring.")

    # ------------------------
    # Low Level I2C Functions
    # ------------------------

    def _write_register(self, reg, data):
        self.i2c.writeto_mem(self.address, reg, data)

    def _read_register(self, reg, length):
        return self.i2c.readfrom_mem(self.address, reg, length)

    def _read_device_id(self):
        data = self._read_register(_REGISTER_VERSION, 2)
        # Device ID is upper byte
        return data[0]

    # ------------------------
    # Temperature Properties
    # ------------------------

    @property
    def temperature(self):
        return self._read_temp(_REGISTER_HOT_JUNCTION)

    @property
    def ambient_temperature(self):
        return self._read_temp(_REGISTER_COLD_JUNCTION)

    @property
    def delta_temperature(self):
        return self._read_temp(_REGISTER_DELTA_TEMP)

    def _read_temp(self, reg):
        data = self._read_register(reg, 2)

        raw = struct.unpack(">h", data)[0]  # signed 16-bit
        return raw * 0.0625

    @property
    def version(self):
        data = self._read_register(_REGISTER_VERSION, 2)
        return struct.unpack(">H", data)[0]

File: C++_files/src/mcp9600.cpp
#include "mcp9600.hpp"
#include "pico/stdlib.h"
#include <cstdio>
#include <stdexcept>
#include <algorithm>

MCP9600::MCP9600(i2c_inst_t* i2c_inst, uint8_t address, TCType type, uint8_t tc_filter)
    : i2c(i2c_inst), addr(address)
{
    tc_filter = std::min<uint8_t>(tc_filter, 7);
    uint8_t ttype = static_cast<uint8_t>(type);
    uint8_t config = tc_filter | (ttype << 4);
    write_register(REG_THERM_CFG, config);

    // Verify device ID (upper byte of version register must be 0x40 or 0x41)
    uint8_t reg = REG_VERSION;
    uint8_t buf[2] = {0};
    i2c_write_blocking(i2c, addr, &reg, 1, true);
    i2c_read_blocking(i2c, addr, buf, 2, false);
    uint8_t device_id = buf[0];
    if (device_id != 0x40 && device_id != 0x41) {
        // In embedded context we print and halt rather than throwing
        printf("FATAL: MCP9600 not found (id=0x%02X). Check wiring.\n", device_id);
        while (true) { tight_loop_contents(); }
    }
}

void MCP9600::write_register(uint8_t reg, uint8_t value) {
    uint8_t buf[2] = {reg, value};
    i2c_write_blocking(i2c, addr, buf, 2, false);
}

int16_t MCP9600::read_raw16(uint8_t reg) {
    uint8_t buf[2] = {0};
    i2c_write_blocking(i2c, addr, &reg, 1, true);
    i2c_read_blocking(i2c, addr, buf, 2, false);
    // Big-endian signed 16-bit
    return (int16_t)((buf[0] << 8) | buf[1]);
}

float MCP9600::read_temp(uint8_t reg) {
    return read_raw16(reg) * 0.0625f;
}

float MCP9600::temperature() {
    return read_temp(REG_HOT_JUNCTION);
}

float MCP9600::ambient_temperature() {
    return read_temp(REG_COLD_JUNCTION);
}

float MCP9600::delta_temperature() {
    return read_temp(REG_DELTA_TEMP);
}

uint16_t MCP9600::version() {
    uint8_t reg = REG_VERSION;
    uint8_t buf[2] = {0};
    i2c_write_blocking(i2c, addr, &reg, 1, true);
    i2c_read_blocking(i2c, addr, buf, 2, false);
    return (uint16_t)((buf[0] << 8) | buf[1]);
}
